Adding error message and placeholder for dropdown

aasimkhan30 opened the pull request:
1. Placeholders in dropdowns is a request from Migration team for adding dropdown instructions in a grid
2. Having dropdown specific error messages will help us having a targeted error message like the azure portal.  

Placeholder for Editable dropdown:
![image](https://user-images.githubusercontent.com/6816294/132376278-73d9116b-d0b8-46fb-8a92-8c4f8d8fb049.png)


Intital state of the dropdown.
![image](https://user-images.githubusercontent.com/6816294/132246635-3b3eb6aa-7511-43b0-b710-976c6b51a416.png)

After user has interacted with the dropdown (type/select an option) once, the dropdown becomes invalid.
![image](https://user-images.githubusercontent.com/6816294/132246687-c1066dae-38fd-4143-9a95-6498fb730eb2.png)

Users can set custom validation messages based on different conditions.
![image](https://user-images.githubusercontent.com/6816294/132268785-06eeedbe-68b6-4a9e-957e-f6b985f3fb0d.png)
![image](https://user-images.githubusercontent.com/6816294/132268870-354dbe31-7dd4-43eb-9474-8e4c8e4a0b5d.png)

Multiple validation errors:

![image](https://user-images.githubusercontent.com/6816294/132620878-bfca6333-cf62-4eec-994d-4b4f92bab906.png)

Validation errors when there is the dropdown has a width
![image](https://user-images.githubusercontent.com/6816294/132621048-6979d31a-ad7c-4237-b6bf-c9470090dace.png)

Charles-Gagnon commented:
Please attach images for all UI changes.

And if you can - please give a brief explanation of why this addition is necessary.

Charles-Gagnon commented on src/sql/azdata.proposed.d.ts:
```diff
@@ -430,6 +430,17 @@ declare module 'azdata' {
+		/**
+		 * Adds a short hint that describes the expected value for the editable dropdown
+		 */
+		editableDropdownPlaceholder?: string;
```
I think just `placeholder` would be better - and then the comment can indicate that it's only used for editable dropdowns.

Charles-Gagnon commented on src/sql/base/parts/editableDropdown/browser/dropdown.ts:
```diff
@@ -344,8 +344,8 @@ export class Dropdown extends Disposable implements IListVirtualDelegate<string>
 	}
 
 	public set value(val: string) {
+		this._input.value = val;
```
Why are you moving this? It seems like the underlying inputbox is also doing this check so I guess it should be fine, but is there a reason you needed to do this in the first place?

aasimkhan30 replied:
There was a bug in editable dropdown that does not show the currently selected option when we accidentally type over a selected option and then select the same option again.  This line of code fixes that change.

Charles-Gagnon replied:
Is this a current issue? In the future it would be good to submit fixes like this in a separate PR. At the very least make sure you explicitly call out that you're fixing an issue and describe what the issue.

aasimkhan30 replied:
Yes. It is an issue in the current state of the dropdown. I will avoid clubbing issue fixes like these in the future with feature PRs

Charles-Gagnon commented on src/sql/workbench/api/common/extHostModelView.ts:
```diff
@@ -1550,6 +1550,22 @@ class DropDownWrapper extends ComponentWrapper implements azdata.DropDownCompone
 		return emitter && emitter.event;
 	}
+
+	public get editableDropdownPlaceholder(): string {
```
`string | undefined`

Charles-Gagnon commented on src/sql/workbench/api/common/extHostModelView.ts:
```diff
@@ -1550,6 +1550,22 @@ class DropDownWrapper extends ComponentWrapper implements azdata.DropDownCompone
+		this.setProperty('editableDropdownPlaceholder', v);
+	}
+
+	public get validationErrorMessage(): string {
```
`string | undefined`

Charles-Gagnon commented on src/sql/workbench/browser/modelComponents/dropdown.component.ts:
```diff
@@ -121,6 +135,13 @@ export default class DropDownComponent extends ComponentBase<azdata.DropDownProp
 
+	public override async validate(): Promise<boolean> {
+		const validationResult = await super.validate();
+		this.isValid = validationResult;
```
The base class already has a valid property - can you not just use that?

Charles-Gagnon commented on src/sql/workbench/browser/modelComponents/dropdown.component.ts:
```diff
@@ -300,4 +323,27 @@ export default class DropDownComponent extends ComponentBase<azdata.DropDownProp
 		});
 	}
+
+	public get editableDropdownPlaceholder(): string {
```
Any reason you're defaulting to an empty string instead of leaving it as undefined? Undefined would be better to use (clearer) for when the user specifically doesn't want a placeholder

Charles-Gagnon commented on src/sql/workbench/browser/modelComponents/dropdown.component.ts:
```diff
@@ -300,4 +323,27 @@ export default class DropDownComponent extends ComponentBase<azdata.DropDownProp
+		return this.getPropertyOrDefault<string>((props) => props.editableDropdownPlaceholder, '');
+	}
+
+	public get validationErrorMessage(): string {
```
Did you verify that this is handled by screen readers correctly? Announced properly and in a way that lets users understand what the message is referring to?

Charles-Gagnon commented on src/sql/workbench/browser/modelComponents/media/dropdown.css:
```diff
@@ -0,0 +1,25 @@
+}
+
+.dropdown-error-icon {
+	background-image: url('status_error.svg');
```
We already have status_error in the common-icons, it would be preferable to add the class to the element and re-use that instead of duplicating the svg.

alanrenmsft commented on src/sql/workbench/browser/modelComponents/dropdown.component.ts:
```diff
@@ -45,6 +52,11 @@ export default class DropDownComponent extends ComponentBase<azdata.DropDownProp
+	/**
+	 * This flag is used to hide the error message in the initial state of the dropdown. We do not want to show the error message before user has interacted with it.
+	 */
+	public isInitState: boolean = true;
```
do you really need this flag? if the default value for isValid is true then you don't need this, right?

aasimkhan30 replied:
The validation is run when the dropdown is first initialized. If the dropdown is empty, the validation will fail, and error will be shown

alanrenmsft commented on src/sql/workbench/browser/modelComponents/dropdown.component.ts:
```diff
@@ -35,6 +38,10 @@ import { ILogService } from 'vs/platform/log/common/log';
 		</div>
 		<div [style.display]="getEditableDisplay()" #editableDropDown style="width: 100%;"></div>
 		<div [style.display]="getNotEditableDisplay()" #dropDown style="width: 100%;"></div>
+		<div class="dropdown-error-container" #errorMessage *ngIf="!isValid && validationErrorMessage.length!==0 && !isInitState">
```
is this approved by the UX team?

aasimkhan30 replied:
I was using this design pattern from the portal. Do I still need to get the UX approval?

aasimkhan30 commented:
@alanrenmsft and @Charles-Gagnon. I have fixed the screen reader issues and added support for multiple validation errors. Currently, it behaves the same way as azure portal dropdown error messages.

alanrenmsft commented on src/sql/workbench/browser/modelComponents/dropdown.component.ts:
```diff
@@ -121,6 +140,33 @@ export default class DropDownComponent extends ComponentBase<azdata.DropDownProp
+		const validationResult = await super.validate();
+		this._changeRef.detectChanges();
+		if (!validationResult) {
+			if (this.editable) {
```
this can be simplified if you define an element variable and then do add attribute or remove attribute

```
const element = this.editable ? this._editableDropdown: this._selectBox.selectElem;
```

aasimkhan30 commented:
Adding border colors for validation errors: 
![image](https://user-images.githubusercontent.com/6816294/132643882-9b7f430f-02b1-476a-825c-47bd96ef528e.png)

Adding focus offset so that we can see the red border. 
![image](https://user-images.githubusercontent.com/6816294/132644172-9084099a-cc0d-4d87-8619-09da49284259.png)

Charles-Gagnon commented on src/sql/workbench/browser/modelComponents/dropdown.component.ts:
```diff
@@ -35,6 +38,16 @@ import { ILogService } from 'vs/platform/log/common/log';
 		</div>
 		<div [style.display]="getEditableDisplay()" #editableDropDown style="width: 100%;"></div>
 		<div [style.display]="getNotEditableDisplay()" #dropDown style="width: 100%;"></div>
+		<label #errorMessage tabindex="-1"
```
Why the tabindex of -1? This doesn't seem like something that would be focusable regardless.

aasimkhan30 replied:
Without it the screen reader was not reading the error message.

aasimkhan30 commented on src/sql/workbench/browser/modelComponents/dropdown.component.ts:
```diff
@@ -35,6 +38,16 @@ import { ILogService } from 'vs/platform/log/common/log';
 		<div [style.display]="getNotEditableDisplay()" #dropDown style="width: 100%;"></div>
+		<label #errorMessage tabindex="-1"
+		aria-live="polite" [attr.id]="errorId" aria-atomic="true"
+		*ngIf="!_valid && validationErrorMessages && validationErrorMessages.length!==0 && !isInitState">
```
![image](https://user-images.githubusercontent.com/6816294/132769903-1329b877-1f2f-4d0c-98b2-6b766732e749.png)

aasimkhan30 replied:
Generally speaking, there shouldn't be multiple validation messages. However, it is better to support them now and force everyone to use arrays instead of adding them later with back compat cases everywhere.

aasimkhan30 commented on src/sql/workbench/browser/modelComponents/dropdown.component.ts:
```diff
@@ -35,6 +38,16 @@ import { ILogService } from 'vs/platform/log/common/log';
+		aria-live="polite" [attr.id]="errorId" aria-atomic="true"
+		*ngIf="!_valid && validationErrorMessages && validationErrorMessages.length!==0 && !isInitState">
+			<ng-container *ngFor="let error of validationErrorMessages">
+				<div  class="dropdown-error-container">
```
When width is not specified.
![image](https://user-images.githubusercontent.com/6816294/132770268-58ba8982-f704-44ec-9ff6-48f97f4bb6dc.png)

When width is specified.
![image](https://user-images.githubusercontent.com/6816294/132770515-5deba75e-8398-4898-bd74-6e5c9c4b4ddb.png)

Charles-Gagnon commented on src/sql/workbench/browser/modelComponents/dropdown.component.ts:
```diff
@@ -35,6 +38,16 @@ import { ILogService } from 'vs/platform/log/common/log';
+		*ngIf="!_valid && validationErrorMessages && validationErrorMessages.length!==0 && !isInitState">
+			<ng-container *ngFor="let error of validationErrorMessages">
+				<div  class="dropdown-error-container">
+					<div class="sql codicon error dropdown-error-icon"></div>
```
What is the sql class for? I don't recall seeing rules for icons that specify anything about a sql class - if you remove it what happens?

aasimkhan30 replied:
Without sql class
![image](https://user-images.githubusercontent.com/6816294/132897425-362add02-5371-4ccc-b2c0-f64a4b6e53a5.png)
With sql class
![image](https://user-images.githubusercontent.com/6816294/132897502-72ace510-1224-47be-bb64-dd6e5915ea90.png)

Charles-Gagnon replied:
Interesting, can you find out where the rule that defines the icon for the sql one is coming from? The inspector should tell you what css file it's in and what the rule selector is.

aasimkhan30 replied:
I got the sql class form the common-icons 

```
.vs .sql.codicon.error,
.vs-dark .sql.codicon.error,
.hc-black .sql.codicon.error {
	content: url("status_error.svg");
}
```

Charles-Gagnon replied:
There we go, thanks. Guess we kept that since the VS Code one isn't filled in so we had our own custom (would be nice to have named it liked error-filled or something..but ya'know...)

Charles-Gagnon commented on src/sql/workbench/browser/modelComponents/dropdown.component.ts:
```diff
@@ -300,4 +344,41 @@ export default class DropDownComponent extends ComponentBase<azdata.DropDownProp
+	if (inputValidationErrorBorderColor) {
+		collector.addRule(`
+		.error-dropdown {
+			border-color: ${inputValidationErrorBorderColor} !important;
```
We should avoid using important when we can - the better way would be to scope the rule down to be more specific.

And generally it's best not to have rules like this be too generic. This rule means that ANYTHING in the app with the class `error-dropdown` is going to have this rule applied, which is not great. 

At the very least you should always be scoping down component rules to just the component, so 

`modelview-dropdown .error-dropdown`

(same for the other rule here)

aasimkhan30 replied:
I am adding important because it is overwriting the styles that are directly added to the component by the base class.

Charles-Gagnon replied:
That's unfortunate (and another thing that wouldn't be an issue if this logic had been done in the base class instead - and/or if the base class used rules instead of applying the style directly)

Charles-Gagnon commented on src/sql/workbench/browser/modelComponents/media/dropdown.css:
```diff
@@ -0,0 +1,24 @@
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+.dropdown-error-container {
```
Rules for components should always be scoped to the components themselves, e.g.

`modelview-dropdown .dropdown-error-container`

(see other css files in this folder for examples)

Charles-Gagnon reviewed (commented):
Should have thought about this earlier, but it would really be best I think to have the error message support be added to the base sql dropdown class itself...by adding the logic here that means that places which don't use the modelview component (such as the connection dialog) won't be able to use this. 

I guess for now this is fine, but in the future please try to do any feature improvements in the base classes and have as little as possible in the angular components since those are much harder to re-use across the app.
